src/main/java/com/pusher/client/channel/impl/PrivateEncryptedChannelImpl.java

```diff
@@ -0,0 +1,152 @@
+import com.pusher.client.channel.PrivateEncryptedChannelEventListener;
+import com.pusher.client.channel.SubscriptionEventListener;
+import com.pusher.client.connection.impl.InternalConnection;
+import com.pusher.client.crypto.nacl.SecretBoxOpener;
+import com.pusher.client.util.Factory;
+import com.pusher.client.util.internal.Base64;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static java.util.Arrays.fill;
+
+public class PrivateEncryptedChannelImpl extends ChannelImpl implements PrivateEncryptedChannel {
+
+    private static final Gson GSON = new Gson();
+    private final InternalConnection connection;
+    private final Authorizer authorizer;
+    private SecretBoxOpener secretBoxOpener;
+
+    private class PrivateEncryptedChannelData {
+        byte[] auth;
+        final String channelData;
+
+        protected PrivateEncryptedChannelData(byte[] auth, String channelData) {
+            this.auth = auth;
+            this.channelData = channelData;
+        }
+
+        public byte[] getAuth() {
+            return auth;
+        }
+
+        public String getChannelData() {
+            return channelData;
+        }
+
+        protected void clearAuthToken() {
+            fill(auth, (byte) 0);
+        }
+    }
+
+    private PrivateEncryptedChannelData authorizerData;
+
+    public PrivateEncryptedChannelImpl(final InternalConnection connection,
+                                       final String channelName,
+                                       final Authorizer authorizer,
+                                       final Factory factory) {
+        super(channelName, factory);
+        this.connection = connection;
+        this.authorizer = authorizer;
+    }
+
+    /* Base class overrides */
+
+    @Override
+    public void bind(final String eventName, final SubscriptionEventListener listener) {
+
+        if (!(listener instanceof PrivateEncryptedChannelEventListener)) {
+            throw new IllegalArgumentException(
+                    "Only instances of PrivateEncryptedChannelEventListener can be bound " +
+                            "to a private encrypted channel");
+        }
+
+        super.bind(eventName, listener);
+    }
+
+
+    protected void checkAuthentication() {
+        final String authResponse = getAuthResponse();
+
+        try {
+            final Map authResponseMap = GSON.fromJson(authResponse, Map.class);
+            final String authKey = (String) authResponseMap.get("auth");
+            final String sharedSecret = (String) authResponseMap.get("shared_secret");
+            final String channelData = (String)authResponseMap.get("channel_data");
+
+            if (authKey == null || sharedSecret == null) {
+                throw new AuthorizationFailureException("Didn't receive all the fields we expected " +
+                        "from the Authorizer, expected an auth token and shared_secret but got: "
+                        + authResponse);
```

Comment:
logging any secrets is not considered good from the perspective of secure coding, likewise putting secrets in exception messages, besides of expanding attack vectors exception messages also end up in a log. I think we should only check `auth` here and throw an exception with more specific message. SecretBoxOpener already does key presence check so will throw the specific exception. passing authResponse to exception message also applies to L98 (I know it's existing issue in PrivateChannelImpl so it feels like it would be good to address that there too).

Reply:
Hmm. That's a really good point. I've removed the two instances in this one where we were logging out the errors.

Reply:
how about the mentioned [PrivateChannelImpl](https://github.com/pusher/pusher-websocket-java/blob/4cdf2b85818b6137fc35f1ae57555fd8cb17151b/src/main/java/com/pusher/client/channel/impl/PrivateChannelImpl.java#L113)?